### PR TITLE
issue180: launch resulting image in container for diagnostic purposes

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees @soltysh Gentlemen, the PR for issue 180.

Fixed https://github.com/openshift/source-to-image/issues/180

### CODE REVIEW

**Comment by bparees** [comment]
@thoraxe also fyi.

File: cmd/sti/main.go [hunk 1]
```diff
@@ -167,10 +168,19 @@ func newCmdBuild(cfg *api.Config) *cobra.Command {
...
+					rerr = runner.Run(cfg)
+				}
+				checkErr(rerr)
+			}
+
 		},
 	}
 
 	buildCmd.Flags().BoolVarP(&(cfg.Quiet), "quiet", "q", false, "Operate quietly. Suppress all non-error output.")
+	buildCmd.Flags().BoolVar(&(cfg.RunImage), "run", false, "Run resulting image as part of invocation of this command.  You may have to Ctrl-C to exit the sit command invocation and the container launched because this options was specified.")
```
Comment (bparees) [inline cmd/sti/main.go]: s/sit/s2i/

Comment (bparees) [inline cmd/sti/main.go]: or maybe sti in this case since you're referring to the actual binary name and we haven't changed it yet.

File: pkg/docker/docker.go [hunk 2]
```diff
@@ -472,17 +479,40 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
...
-	glog.V(2).Infof("Container exited")
+	if opts.TargetImage {
+		exposedPorts := make([]string, 0)
+		for port := range imageMetadata.Config.ExposedPorts {
+			exposedPorts = append(exposedPorts, port.Port())
+		}
 
-	if exitCode != 0 {
-		return errors.NewContainerError(container.Name, exitCode, "")
+		glog.V(0).Infof("\n\n\n\n\nThe image %s has been started in container %s as a result of the --run=true option.  The container's stdout/stderr will be redirected to this command's glog output to help you validate its behavior.  You can also inspect the container with docker commands if you like.  If the container is set up to stay running, you will have to Ctrl-C to exit this command, which should also stop the container %s.  This particular invocation attempts to run on the exposed ports %+v \n\n\n\n\n", image, container.ID, container.ID, exposedPorts)
```
Comment (bparees) [inline pkg/docker/docker.go]: I was thinking you'd print the port mappings (ie the host port and what container port it mapped to).  Just the internal port numbers aren't too useful here.

File: pkg/docker/docker.go [hunk 3]
```diff
@@ -352,39 +355,43 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
...
-		commandBaseDir = filepath.Join(tarDestination, "scripts")
-		glog.V(2).Infof("Both scripts and untarred source will be placed in '%s'", tarDestination)
-	} else {
-		// for internal scripts we can have separate path for scripts and untar operation destination
-		scriptsURL := opts.ScriptsURL
-		if len(scriptsURL) == 0 {
-			scriptsURL = getScriptsURL(imageMetadata)
+
+	tarDestination := ""
+	if len(opts.Command) != 0 {
```
Comment (bparees) [inline pkg/docker/docker.go]: i wouldn't overload the meaning of an empty opts.Command to mean "running the target image" because we might in the future want to support passing a command to run in the target image.  ie --run --command=<somecommand> instead of only running the default image CMD/ENTRYPOINT.

File: cmd/sti/main.go (hunk 1 above)
Comment (gabemontero) [inline cmd/sti/main.go]: that's right - used `sti` cause referring to binary

File: pkg/docker/docker.go (hunk 3 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: yep - good catch - that may have been a leftover from the iterations / permutations around the sleep/wait/signal stuff and how the `sti` command exited.  Switching to a check of the opts.TargetImage boolean.

File: pkg/docker/docker.go (hunk 2 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: OK, so I *think* this means in the Ports map at https://github.com/openshift/source-to-image/blob/master/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/container.go#L131
as the PortMapping map above seems to be in reference to some deprecated `docker inspect` output.

Network settings is off of the Container struct:  https://github.com/openshift/source-to-image/blob/master/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/container.go#L221-L235

And we get a ref to the container earlier in this method.

So yep, I'll switch it and give it a whirl.

**Comment by thoraxe** [comment]
@bparees this is way over my head...

File: cmd/sti/main.go (hunk 1 above)
Comment (bparees) [inline cmd/sti/main.go]: still needs to be fixed from sit to sti

**Comment by bparees** [comment]
@thoraxe i just wanted you to see the trouble you cause when you ask for stuff :)

File: pkg/docker/docker.go (hunk 2 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: Hmm ... unfortunately, the NetworkSettings are coming up `nil` in my testing.  Perhaps the Container ref returned by CreateContainer is not updated when other operations like StartContainer are done later on.

I could look into exposing https://github.com/openshift/source-to-image/blob/master/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/container.go#L57-L69, which if called after StartContainer, perhaps the APIContainer / APIPort at https://github.com/openshift/source-to-image/blob/master/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/container.go#L32-L52 could get us what you want.

Comment (gabemontero) [inline pkg/docker/docker.go]: Yep, that seems to have done the trick.  Will be part of upcoming push.

**Comment by gabemontero** [comment]
ok ... push with updates from comments done

File: cmd/sti/main.go (hunk 1 above)
Comment (bparees) [inline cmd/sti/main.go]: still says "sit" and not "sti"

File: pkg/docker/docker.go [hunk 4]
```diff
@@ -472,17 +480,50 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
...
-	}
-	glog.V(2).Infof("Container exited")
+	if opts.TargetImage {
 
-	if exitCode != 0 {
-		return errors.NewContainerError(container.Name, exitCode, "")
+		lco := docker.ListContainersOptions{
+			All: true,
+		}
+		conts, lcerr := d.client.ListContainers(lco)
```
Comment (bparees) [inline pkg/docker/docker.go]: the client api doesn't give you a way to just get the container given that you've got the container id?  weird.

File: pkg/docker/docker.go [hunk 5]
```diff
@@ -472,17 +480,50 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
...
+		if lcerr == nil {
+			for _, cont := range conts {
+				if cont.ID == container.ID {
+					liveports = cont.Ports
+					break
+				}
+			}
+		}
+
+		glog.V(0).Infof("\n\n\n\n\nThe image %s has been started in container %s as a result of the --run=true option.  The container's stdout/stderr will be redirected to this command's glog output to help you validate its behavior.  You can also inspect the container with docker commands if you like.  If the container is set up to stay running, you will have to Ctrl-C to exit this command, which should also stop the container %s.  This particular invocation attempts to run with the port mappings %+v \n\n\n\n\n", image, container.ID, container.ID, liveports)
```
Comment (bparees) [inline pkg/docker/docker.go]: can you show us a sample of this output?  i'm curious in particular what the port data looks like

**Comment by bparees** [comment]
one nit and a couple Qs and then i think this lgtm, pending what the example output looks like.

i'd like to get @soltysh's sign off too though since he's most familiar w/ some of this code and might be aware of gotchas.

File: cmd/sti/main.go (hunk 1 above)
Comment (gabemontero) [inline cmd/sti/main.go]: Sorry, was distracted with the other comments, missed this one.  Part of next push.

File: pkg/docker/docker.go (hunk 4 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: Actually I just found an InspectContainer.  Going to switch to that one, assuming it works.

Comment (gabemontero) [inline pkg/docker/docker.go]: Nope - InspectContainer and the Container / NetworkSettings / Port / PortBinding only lists:

map[8080/tcp:[]]

The ListContainer / APIContainer / APIPort has more info.  Staying with that.

File: pkg/docker/docker.go (hunk 5 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: Here you go:

 [{PrivatePort:8080 PublicPort:0 Type:tcp IP:}] 

While it has an IP field, it is showing up empty, perhaps because we are running against the local docker daemon.  Of course public port 0 is not super informative either.  And cross referencing with netstat -nap was not conclusive.

Don't think we are going to get much better than this though.  I'll push, but I'm thinking this is not worth it, and the user will just want to run docker cmds, netstat, etc. on his own.

Comment (bparees) [inline pkg/docker/docker.go]: are you sure the ports *are* getting mapped to something on the host?
(easy test:  run s2i with --run and then do a docker ps
here's an example of a container i ran with "docker run -P" so the exposed ports got mapped:
$ docker ps
CONTAINER ID        IMAGE                       COMMAND             CREATED             STATUS              PORTS                     NAMES
961ab6e2203a        openshift/ruby-20-centos7   "/bin/sh"           26 seconds ago      Up 23 seconds       0.0.0.0:32768->8080/tcp   insane_hodgkin

File: pkg/build/strategies/sti/sti.go [hunk 6]
```diff
@@ -403,7 +398,7 @@ func (b *STI) Execute(command string, config *api.Config) error {
 		}
 	}(outReader)
 
-	go streamContainerError(errReader, &errOutput, config)
+	go docker.StreamContainerIO(errReader, &errOutput, glog.Error)
```
Comment (soltysh) [inline pkg/build/strategies/sti/sti.go]: :8ball:

File: pkg/build/interfaces.go [hunk 7]
```diff
@@ -49,6 +49,12 @@ type Ignorer interface {
 	Ignore(*api.Config) error
 }
 
+// Runner allows the user to actually run the produced imgae as part of the s2i
+// cmd line invocation, to see if it operates as he would expect
+type Runner interface {
+	Run(*api.Config) error
+}
```
Comment (soltysh) [inline pkg/build/interfaces.go]: There's no need to add this interface here. First of all you're not using it anywhere inside the build. Moreover it is not related to build in any way.

File: pkg/docker/docker.go [hunk 8]
```diff
@@ -472,17 +480,50 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
...
+		if err != nil {
+			return err
+		}
+
+		glog.V(2).Infof("Container exited")
+		if exitCode != 0 {
+			return errors.NewContainerError(container.Name, exitCode, "")
+		}
 	}
+
```
Comment (soltysh) [inline pkg/docker/docker.go]: I'd suggest splitting this function into smaller parts, it's almost 200 lines long and it's damn hard to follow efficiently, my heads explodes trying to follow the logic in it. Is it possible somehow?

File: pkg/docker/docker.go (hunk 5 above)
Comment (soltysh) [inline pkg/docker/docker.go]: Just checked it and I have to complain :disappointed: because the ports are NOT mapped. I can't access localhost:8080 after doing: 
```
sti build https://github.com/openshift/sti-python.git --run --loglevel=3 --context-dir=3.4/test/setup-test-app/ openshift/python-34-centos7 python-sample-app
```
Running `docker run -p 8080:8080 python-sample-app` allows me to access the ports.

**Comment by soltysh** [comment]
Good work @gabemontero, I'll reserve some more time later today or at latest tomorrow morning to analyze it more thoroughly.

File: pkg/docker/docker.go (hunk 8 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: Yeah, it was long  before I touched it ;-), but certainly my change only added to the situation.  I'll see what I can do.

**Comment by gabemontero** [comment]
Thanks @soltysh .

File: pkg/docker/docker.go (hunk 8 above)
Comment (soltysh) [inline pkg/docker/docker.go]: I know it was long :stuck_out_tongue_winking_eye: but now I found someone lucky enough to fix it :wink: 
But srsly, I've looked at simplifying it a couple of times before, but frankly said I failed, it's too tightly coupled.

File: pkg/build/interfaces.go (hunk 7 above)
Comment (gabemontero) [inline pkg/build/interfaces.go]: Don't see why being used inside the build or related to the build matters wrt leveraging an interface.  I see the interfaces as a way to outline and define the semantics of major portions of the entire program - concepts that are of interest to the user.

File: pkg/docker/docker.go (hunk 5 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: I suspect I'm about to receive some more education on working in the docker world ;-) ... so after looking at `docker help run`, I see that `-p` publishes specific ports, and -P publishes all ports ... assuming that is the mapping you all are referring to, then it seems I need to look at the docker client api and see what triggers the equivalent of -p/-P ...

... do I sound anywhere close ?

File: pkg/docker/docker.go (hunk 8 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: LOL - sounds good - it may take a couple of iterations of push/review on our respective parts, but let's see where we land.

File: pkg/build/interfaces.go (hunk 7 above)
Comment (soltysh) [inline pkg/build/interfaces.go]: Yes, but these interfaces specifically are for builds only (pkg/**build**/interfaces.go). And as long as providing an interface makes sense here you have only one implementation of that interface and you're not using the interface directly anywhere, so generally speaking you don't leverage the power of that interface.

File: pkg/docker/docker.go (hunk 5 above)
Comment (soltysh) [inline pkg/docker/docker.go]: I'm completely fine with publishing all the ports (`-P`), it's just I've been always using `-p`, so it's a matter of taste.

File: pkg/docker/docker.go (hunk 8 above)
Comment (soltysh) [inline pkg/docker/docker.go]: :+1:

File: pkg/build/interfaces.go (hunk 7 above)
Comment (gabemontero) [inline pkg/build/interfaces.go]: OK, I see the sub directory point.  It could be moved to an interfaces.go file in another directory.

Use of interfaces can be warranted not just for the current situation, but for the anticipation of needing future implementations in the future.  And if you defer on creating the interface until that need arises, your *initial* implementation often needs some rework to keep it in line with the implicit semantics, whereas explicitly defining them helps scope the situation.

Comment (soltysh) [inline pkg/build/interfaces.go]: There's a bigger plan to abstract the entire flow of s2i build to remove unnecessary code duplication we have in origin. Let's first do that and then think about adding additional interfaces if you don't mind. Besides OS will never leverage that Runner interface in the first place, it's solely for s2i standalone edition.

Comment (gabemontero) [inline pkg/build/interfaces.go]: That is true (abstract / rework the existing flow) ... OK, I'll acquiesce here.

File: pkg/docker/docker.go (hunk 5 above)
Comment (bparees) [inline pkg/docker/docker.go]: @gabemontero yes that sounds about right.  The other difference is that with "-p" you can specify the exact mapping (ie "map port 1234 on my host to port 8080 in the container"

with -P docker just allocates host ports randomly.   that's good and bad...  good:  it will (presumably) not pick ports that are already in use by something else.  bad: you have no idea what host port got assigned for a given container port, so you need to go look up the mappings if you want to, for example, access your app via your browser.

that last point is why, if we're going to do -P equivalence, we need to print out the mappings that got assigned, so the user doesn't have to go look it up themselves.

Comment (gabemontero) [inline pkg/docker/docker.go]: OK, I've got it working.  Now, there are two API's /structs that are accessible from the docker client that display the mappings, each in a slightly different format:

list containers -> APIPort
[{PrivatePort:8080 PublicPort:32768 Type:tcp IP:0.0.0.0}]

inspect container -> Port
map[8080/tcp:[{HostIP:0.0.0.0 HostPort:32769}]] 

I've bounced back and forth a bit on this one, but I think I like the inspect container.

Thoughts?

Comment (soltysh) [inline pkg/docker/docker.go]: Doesn't matter whichever you choose, as long as we list those information to user clearly, imho.

Comment (bparees) [inline pkg/docker/docker.go]: +1.  i guess i agree the 2nd one is a little better.  you could of course print out the fields yourself too and format it however you want.

Comment (soltysh) [inline pkg/docker/docker.go]: That's what I meant with my comment, above is not readable imho. I'd rather see line by line with a port mapping information, clearly standing out from the rest of the text.

Comment (gabemontero) [inline pkg/docker/docker.go]: Yeah, I was trying to get the formatting for free.  I'll create a formatted string.

Comment (gabemontero) [inline pkg/docker/docker.go]: Example of the new format:
Port Bindings:  
  Private Port:  8080
        Protocol:  tcp
        Public Host / Port Mappings:
            IP: 0.0.0.0 Port: 32770

Comment (gabemontero) [inline pkg/docker/docker.go]: note - github comments is removing indentation that I added for the various lines ... private port has 1 level of indentation, protocol and mappings header 2 levels, the ip/ port 3 levels

Comment (bparees) [inline pkg/docker/docker.go]: i'd probably label Private Port as Container Port instead, otherwise lgtm.

**Comment by gabemontero** [comment]
OK ... push done with changes in response to latest @bparees comments and initial set of @soltysh comments

File: docs/cli.md [hunk 9]
```diff
@@ -78,6 +78,7 @@ that image and add them to the tar streamed to the container into `/artifacts`.
 | `-e (--env)`               | Environment variables to be passed to the builder eg. `NAME=VALUE,NAME2=VALUE2,...` |
 | `-E (--environment-file)`               | Specify the path to the file with environment |
 | `--force-pull`             | Always pull the builder image, even if it is present locally (defaults to true) |
+| `--run`             | Launch the resulting image in a container, where stdout and stderr from the container are piped to the stdout and stderr of sti to help determine the image's validity (along with any docker commands you which to execute on the container provided), and where if the image is long running, you will have to Ctrl-C to exit both sti and the container (defaults to false) |
```
Comment (soltysh) [inline docs/cli.md]: I'd suggest to rewrite it like this:
Launch the resulting image after a successful build. All output from the image is being printed to help determine image's validity. In case of a long running image you will have to Ctrl-C to exit both sti and the running container.

File: pkg/docker/docker.go [hunk 10]
```diff
@@ -336,6 +340,142 @@ func getDestination(image *docker.Image) string {
 	return DefaultDestination
 }
 
+func runContainerTar(opts RunContainerOptions, config docker.Config, imageMetadata *docker.Image) (docker.Config, string) {
+	tarDestination := ""
+	if !opts.TargetImage {
```
Comment (soltysh) [inline pkg/docker/docker.go]: Fail fast, iow:
```
if opts.TargetImage {
    return config, tarDestination
}
// rest of the code but this time without the indentation 
```

File: pkg/docker/docker.go [hunk 11]
```diff
@@ -336,6 +340,142 @@ func getDestination(image *docker.Image) string {
...
+			liveports = liveports + "\n        Protocol:  " + port.Proto()
+			liveports = liveports + "\n        Public Host / Port Mappings:"
+			for _, binding := range bindings {
+				liveports = liveports + "\n            IP: " + binding.HostIP + " Port: " + binding.HostPort
+			}
+		}
+		liveports = liveports + "\n"
+	}
+
+	glog.V(0).Infof("\n\n\n\n\nThe image %s has been started in container %s as a result of the --run=true option.  The container's stdout/stderr will be redirected to this command's glog output to help you validate its behavior.  You can also inspect the container with docker commands if you like.  If the container is set up to stay running, you will have to Ctrl-C to exit this command, which should also stop the container %s.  This particular invocation attempts to run with the port mappings %+v \n\n\n\n\n", image, container.ID, container.ID, liveports)
```
Comment (soltysh) [inline pkg/docker/docker.go]: No need for V(0), just glog.Infof.

File: pkg/run/run.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		glog.Errorf("Create of Docker Container client failed with %v \n", ccerr)
+		return nil, ccerr
+	}
+	dr := &DockerRunner{
+		ContainerClient: cc,
+	}
+	return dr, nil
+}
+
+func (b *DockerRunner) Run(config *api.Config) error {
```
Comment (soltysh) [inline pkg/run/run.go]: Missing godoc.

File: pkg/run/run.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/openshift/source-to-image/pkg/docker"
+	"github.com/openshift/source-to-image/pkg/errors"
+	"io"
+)
+
+type DockerRunner struct {
+	ContainerClient docker.Docker
+}
+
+func New(config *api.Config) (*DockerRunner, error) {
```
Comment (soltysh) [inline pkg/run/run.go]: Missing godoc.

File: pkg/run/run.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"github.com/golang/glog"
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/docker"
+	"github.com/openshift/source-to-image/pkg/errors"
+	"io"
+)
+
+type DockerRunner struct {
```
Comment (soltysh) [inline pkg/run/run.go]: Missing godoc.

File: pkg/docker/util.go [hunk 15]
```diff
@@ -38,6 +43,25 @@ func GetImageRegistryAuth(dockerCfg io.Reader, imageName string) client.AuthConf
 	return client.AuthConfiguration{}
 }
 
+func StreamContainerIO(errStream io.Reader, errOutput *string, log func(...interface{})) {
```
Comment (soltysh) [inline pkg/docker/util.go]: Missing godoc.

File: pkg/docker/docker.go [hunk 16]
```diff
@@ -336,6 +340,142 @@ func getDestination(image *docker.Image) string {
 	return DefaultDestination
 }
 
+func runContainerTar(opts RunContainerOptions, config docker.Config, imageMetadata *docker.Image) (docker.Config, string) {
```
Comment (soltysh) [inline pkg/docker/docker.go]: godoc all those new functions, to avoid future problems understanding the flow.

**Comment by soltysh** [comment]
I like it a lot. Thanks for splitting that method, reads better now! A couple of small nits, mostly missing godocs, otherwise LGTM.

File: docs/cli.md (hunk 9 above)
Comment (gabemontero) [inline docs/cli.md]: in next push

File: pkg/docker/docker.go (hunk 10 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: in next push

File: pkg/docker/docker.go (hunk 11 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: in next push

File: pkg/run/run.go (hunk 12 above)
Comment (gabemontero) [inline pkg/run/run.go]: in next push

File: pkg/run/run.go (hunk 13 above)
Comment (gabemontero) [inline pkg/run/run.go]: in next push

File: pkg/run/run.go (hunk 14 above)
Comment (gabemontero) [inline pkg/run/run.go]: in next push

File: pkg/docker/util.go (hunk 15 above)
Comment (gabemontero) [inline pkg/docker/util.go]: in next push

File: pkg/docker/docker.go (hunk 16 above)
Comment (gabemontero) [inline pkg/docker/docker.go]: in next push

**Comment by gabemontero** [comment]
Cool deal - ok, nits, comments/godocs pushed.

**Comment by bparees** [comment]
lgtm [merge]

File: pkg/docker/util.go [hunk 17]
```diff
@@ -38,6 +43,27 @@ func GetImageRegistryAuth(dockerCfg io.Reader, imageName string) client.AuthConf
 	return client.AuthConfiguration{}
 }
 
+// Takes data from the Reader and redirects to the log functin (typically we pass in
+// glog.Error for stderr and glog.Info for stdout
+func StreamContainerIO(errStream io.Reader, errOutput *string, log func(...interface{})) {
```
Comment (soltysh) [inline pkg/docker/util.go]: For future, godoc needs to start with the method name you're describing:
```
// StreamContainerIO takes data from the Reader and redirects to the log functin (typically we pass in
// glog.Error for stderr and glog.Info for stdout
func StreamContainerIO(errStream io.Reader, errOutput *string, log func(...interface{})) {
```

File: pkg/run/run.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+	//NOTE, we've seen some Golang level deadlock issues with the streaming of cmd output to
+	// glog, but part of the deadlock seems to have occurred when stdout was "silent"
+	// and produced no data, such as when we would do a git clone with the --quiet option.
+	// We have not seen the hang when the Cmd produces output to stdout.
+
+	go docker.StreamContainerIO(errReader, nil, glog.Error)
+	go docker.StreamContainerIO(outReader, nil, glog.Info)
+	rerr := b.ContainerClient.RunContainer(opts)
+	if e, ok := rerr.(errors.ContainerError); ok {
+		return errors.NewContainerError(config.Tag, e.ErrorCode, errOutput)
```
Comment (rhcarvalho) [inline pkg/run/run.go]: @gabemontero I'm trying to understand this, why do we replace a `ContainerError` with another `ContainerError`?

Comment (gabemontero) [inline pkg/run/run.go]: @rhcarvalho we are updating the contents of the message ... if I recall correctly, the original contents were not as insightful
